File: app/pipelines/podcasts.py
"""Podcast episodes pipeline using the iTunes Search API."""

import logging

import requests

from app.models import db, AlertHistory, TrackedEntity

logger = logging.getLogger(__name__)

ITUNES_SEARCH_URL = "https://itunes.apple.com/search"
ITUNES_LOOKUP_URL = "https://itunes.apple.com/lookup"


def check_podcasts(entity_names, user_id):
    """Check for new podcast episodes for tracked shows.

    Returns a list of new result dicts.
    """
    all_new = []

    for name in entity_names:
        entity = TrackedEntity.query.filter_by(
            user_id=user_id, entity_name=name
        ).join(TrackedEntity.category).filter_by(slug="podcasts").first()

        if not entity:
            continue

        try:
            results = _fetch_episodes(name)
        except Exception:
            logger.exception("Failed to fetch podcast episodes for %s", name)
            continue

        seen_keys = _get_seen_keys(user_id, entity.id)

        for item in results:
            key = _make_key(item)
            if key in seen_keys:
                continue

            alert = AlertHistory(
                user_id=user_id,
                tracked_entity_id=entity.id,
            )
            alert.set_content(item)
            db.session.add(alert)
            all_new.append(item)

    db.session.commit()
    return all_new


def _fetch_episodes(podcast_name):
    """Search iTunes for a podcast and fetch recent episodes."""
    # Search for the podcast
    params = {
        "term": podcast_name,
        "media": "podcast",
        "entity": "podcast",
        "limit": 1,
    }
    resp = requests.get(ITUNES_SEARCH_URL, params=params, timeout=15)
    resp.raise_for_status()
    data = resp.json()

    podcasts = data.get("results", [])
    if not podcasts:
        return []

    collection_id = podcasts[0].get("collectionId")
    if not collection_id:
        return []

    # Lookup episodes
    params = {
        "id": collection_id,
        "media": "podcast",
        "entity": "podcastEpisode",
        "limit": 10,
    }
    resp = requests.get(ITUNES_LOOKUP_URL, params=params, timeout=15)
    resp.raise_for_status()
    data = resp.json()

    results = []
    for item in data.get("results", []):
        if item.get("wrapperType") != "podcastEpisode":
            continue

        results.append({
            "type": "podcast_episode",
            "podcast": podcast_name,
            "episode_title": item.get("trackName", ""),
            "published": item.get("releaseDate", ""),
            "duration_ms": item.get("trackTimeMillis", 0),
            "link": item.get("trackViewUrl", ""),
        })
    return results


def _make_key(item):
    """Create a dedup key from a result item."""
    return f"{item['podcast']}|{item['episode_title']}"


def _get_seen_keys(user_id, entity_id):
    """Get set of already-seen dedup keys for this user+entity."""
    alerts = AlertHistory.query.filter_by(
        user_id=user_id, tracked_entity_id=entity_id
    ).all()
    keys = set()
    for a in alerts:
        content = a.get_content()
        keys.add(_make_key(content))
    return keys

File: packages/shared/src/pipelines/podcasts.ts
import { PipelineContext, PipelineResult } from './types';

const ITUNES_SEARCH_URL = 'https://itunes.apple.com/search';

export async function checkPodcasts(ctx: PipelineContext): Promise<PipelineResult[]> {
  const results: PipelineResult[] = [];

  for (const entity of ctx.entities) {
    const trackMode = (entity.entity_metadata?.track_mode as string) || 'guest';

    try {
      if (trackMode === 'show') {
        const episodes = await fetchShowEpisodes(entity.entity_name);
        for (const ep of episodes) {
          const dedupKey = `show|${entity.entity_name}|${ep.episodeTitle}`;

          results.push({
            entity_name: entity.entity_name,
            tracked_entity_id: entity.id,
            dedup_key: dedupKey,
            event_date: ep.published,
            content: {
              type: 'podcasts',
              track_mode: 'show',
              podcast_name: entity.entity_name,
              episode_title: ep.episodeTitle,
              published: ep.published,
              duration_ms: ep.durationMs,
              url: ep.link,
            },
          });
        }
      } else {
        const episodes = await fetchGuestEpisodes(entity.entity_name);
        for (const ep of episodes) {
          const dedupKey = `guest|${entity.entity_name}|${ep.podcastName}|${ep.episodeTitle}`;

          results.push({
            entity_name: entity.entity_name,
            tracked_entity_id: entity.id,
            dedup_key: dedupKey,
            event_date: ep.published,
            content: {
              type: 'podcasts',
              track_mode: 'guest',
              person: entity.entity_name,
              podcast_name: ep.podcastName,
              episode_title: ep.episodeTitle,
              published: ep.published,
              duration_ms: ep.durationMs,
              url: ep.link,
            },
          });
        }
      }
    } catch (err) {
      console.error(`Podcasts pipeline error for "${entity.entity_name}":`, err);
    }
  }

  return results;
}

interface Episode {
  podcastName: string;
  episodeTitle: string;
  published: string;
  durationMs: number;
  link: string;
}

async function fetchShowEpisodes(podcastName: string): Promise<Episode[]> {
  const params = new URLSearchParams({
    term: podcastName,
    media: 'podcast',
    entity: 'podcastEpisode',
    limit: '20',
  });

  const res = await fetch(`${ITUNES_SEARCH_URL}?${params}`, {
    signal: AbortSignal.timeout(15000),
  });
  if (!res.ok) return [];

  const data = await res.json();
  const episodes: Episode[] = [];
  const nameLower = podcastName.toLowerCase();

  for (const item of data.results || []) {
    if (item.wrapperType !== 'podcastEpisode') continue;

    // Only include episodes from the specific podcast
    const collectionName = (item.collectionName || '') as string;
    if (collectionName.toLowerCase() !== nameLower) continue;

    episodes.push({
      podcastName: collectionName,
      episodeTitle: (item.trackName || '') as string,
      published: item.releaseDate || '',
      durationMs: item.trackTimeMillis || 0,
      link: item.trackViewUrl || '',
    });
  }

  return episodes;
}

async function fetchGuestEpisodes(personName: string): Promise<Episode[]> {
  const params = new URLSearchParams({
    term: personName,
    media: 'podcast',
    entity: 'podcastEpisode',
    limit: '20',
  });

  const res = await fetch(`${ITUNES_SEARCH_URL}?${params}`, {
    signal: AbortSignal.timeout(15000),
  });
  if (!res.ok) return [];

  const data = await res.json();
  const episodes: Episode[] = [];
  const nameLower = personName.toLowerCase();

  for (const item of data.results || []) {
    if (item.wrapperType !== 'podcastEpisode') continue;

    const title = (item.trackName || '') as string;
    const description = (item.description || item.shortDescription || '') as string;

    // Only include episodes where the person's name appears in the title or description
    if (
      !title.toLowerCase().includes(nameLower) &&
      !description.toLowerCase().includes(nameLower)
    ) continue;

    episodes.push({
      podcastName: item.collectionName || '',
      episodeTitle: title,
      published: item.releaseDate || '',
      durationMs: item.trackTimeMillis || 0,
      link: item.trackViewUrl || '',
    });
  }

  return episodes;
}
